### ISSUE TITLE
Better hashing strategy with NumPy arrays

### ISSUE THREAD

**Issue by jakirkham** [dask/dask#4169]
It seems the current hashing strategy of hashing all NumPy array contents has bitten a few users and is generally felt as surprising. Given the number of times `da.asarray` ends up being called to ensure constraints are met, this can be quite surprising for users. Would suggest that we either change the default hashing behavior or just turn it off.

As an initial suggestion about how we can change it, would try using `id` on the NumPy array to inform its hash. This is probably unique enough for most cases and we can be pretty sure that cases where that key shows up the NumPy array still exists in our graph.

Thoughts?

**Comment by jakirkham** [comment]
cc @hmaarrfk (who probably has thoughts here)

**Comment by hmaarrfk** [comment]
I think this comment https://github.com/dask/dask/issues/3946#issuecomment-418568246 from Matt is really relevant.

I feel like you and I are working with people that are "transitionning" to using dask, where the whole pipeline isn't dask-y yet. So we are hitting a few of this hiccups. Matt seems to have a different usecase which he seems to have optimized dask array for.

I just learned to put `name=False`. In either case, I think whatever internal functions are written should use the name parameter in an explicit fashion. This will help the transition if it ever happens.

**Comment by jakirkham** [comment]
Sure. Though am asking we reconsider this feature as it is currently optimized for experienced users who are better able to adapt at the expense of new users to Dask who are just learning.

In any event, would think that basing the hash off of `id(a)` ends up being a reasonable compromise.

**Comment by hmaarrfk** [comment]
well it becomes a matter of creating arrays such as `da.ones`, or `da.zeros` in many places inn the code. Those would get caught by the hash pretty easly, but not by id.

I'm sure other niche things like fourier coefficients would slip through the `id(a)` while be caught by other stuff.

Matt (not using handle as not to ping him needlessly)  will probably object to this since he probably has alot of code in the wild optimized with the assumption that things will get hashed.
You would totally need to deprecate this gracefully. 

I wrote some decorators for this a while back.
https://github.com/hmaarrfk/wabisabi
Don't know if they are useful for you.

They do crazy things like rewrite the docs, the signature, and handle positional + kwargs.
The advantage is you can leave them in past the deprecation, and everything **should** just get updated gracefully. The package name changed, so they are no longer on conda-forge.

You should probably version it if you will use them.

**Comment by jakirkham** [comment]
`da.ones` and friends actually compute their hash [from the arguments to the function]( https://github.com/dask/dask/blob/66f5e0bbb31c503063616610475c50deaf1eccb8/dask/array/wrap.py#L43 ) and not the arrays themselves. So they shouldn't be affected by this proposed change.

FFTs in Dask just [wrap NumPy's implementations]( https://github.com/dask/dask/blob/66f5e0bbb31c503063616610475c50deaf1eccb8/dask/array/fft.py#L216-L229 ). So they should be hashing other keys and thus should be unaffected. Or is there another use case you have in mind?

Anyways thinking about use cases that might be affected seems like the right idea. Would be good to collect some use cases in the wild to see if they would be affected by this change. My feeling ATM is that these use cases are exceedingly rare and that any such use case is not likely reconstructing the same NumPy array multiple times, in which case `id` should be fine. Even if an algorithm were creating the same array repeatedly it would be a straightforward thing to set the `name` explicitly and not rely on hashing. Though I could be wrong about all of this.

Interesting. Thanks for sharing. Will take a look.

**Comment by hmaarrfk** [comment]
> da.ones and friends actually compute their hash from the arguments to the function and not the arrays themselves. So they shouldn't be affected by this proposed change.

So smart!

**Comment by jcrist** [comment]
> As an initial suggestion about how we can change it, would try using id on the NumPy array to inform its hash. This is probably unique enough for most cases and we can be pretty sure that cases where that key shows up the NumPy array still exists in our graph.

Ids in python can be reused, I'd be -1 on any strategy that has the potential to be incorrect just for the benefit of speed.

There is an option to use a `uuid` for wrapped arrays (`name=False` kwarg to `from_array`/`asarray`), which would be similar but more robust. Switching this to the default hashing would likely break users code, but perhaps we could make this part of the configuration. You could then provide a config file to novice users of your software, or set configuration variables on init of your module (or contextually in your code).

> I think this comment https://github.com/dask/dask/issues/3946#issuecomment-418568246 from Matt is really relevant.

In general I agree with Matt's comment that most users of dask array appreciate the current behavior, as hashing large numpy arrays is relatively rare in common workflows.

**Comment by jakirkham** [comment]
Yeah, it sounds like there are different groups of users that have very different expectations about what `from_array`/`asarray` should do. However I don't see a good way of writing code in Dask proper that can correctly handle all of these needs directly. Moving to a configuration option sounds like a good idea to get this option generally into the hands of users as cleanly as possible.

**Comment by jakirkham** [comment]
FWIW I think this remains an issue that people care about. The xref'd issue from the user above was solicited after talking with him about issues he had encountered with Dask (this issue being one of them). It would be good to figure out a strategy to make hashing more customizable.

**Comment by jakirkham** [comment]
Thinking about this again maybe one naive approach would be to use a UUID when arrays are larger than a certain number of bytes. We’d have to determine a sensible cut off, but maybe this captures the best of both worlds? 🙂

**Comment by uellue** [comment]
The discussion in #7355 shows a real-world use case where a Dask array is created from a very large memory-mapped NumPy array. In this particular case the hashing was really surprising and led us off track when trying to find out why one version was much slower than the other. Hashing takes about 20x longer than the actual array operation for that use case and negates any advantage from memory mapping.

About creating Dask arrays from memory: Using Dask to work with very large NumPy arrays could be a legit way to speed up some operations through multithreading and at the same time reduce their memory footprint since large intermediate arrays are eliminated, right?

Are there examples for use cases where the hashing is proven to bring a benefit? I can imagine that it helps mostly in cases where the array is so small that overheads from Dask dominate over the actual number crunching, meaning that the benefit from eliminating steps in the calculation is worth the additional cost of hashing? If the array is larger, many operations such as slicing, sum etc are actually faster than the hash function, so it hurts less to just do some extra number crunching than to avoid it by hashing.

Making a decision based on size sounds sensible for that reason. Maybe about 10 MB could be a good cut-off?

Are there scenarios where hashing or not would lead to different results, i.e. where it is not only a performance issue, but a correctness issue?
